gtdynamics/universal_robot/sdf.cpp

```diff
@@ -29,12 +30,21 @@ using gtsam::Pose3;
 sdf::Model GetSdf(const std::string &sdf_file_path,
                   const std::string &model_name,
                   const sdf::ParserConfig &config) {
-  sdf::SDFPtr sdf = sdf::readFile(sdf_file_path);
-  if (sdf == nullptr)
+  sdf::ParserConfig _config = config;
+  _config.URDFSetPreserveFixedJoint(true);
```

Comment:
I was just thinking about a use case where someone (e.g. Jerred?  idk) doesn't care about the fixed joints and doesn't want the computation penalty of 70% more joints so they'd rather have the SDF do its auto-merging of links.  Could we maybe:
1) have a flag for whether to use the fixed joints or not
2) just remove these 2 lines and if you want to have explicit fixed joints then pass in the appropriate config?  (i.e. make erase-fixed-joints the default)

Because otherwise there's no way to un-preserve the fixed joints.

Reply:
Where is the computational penalty coming from? If it's forward kinematics then fixed joints don't participate in that since they don't have any values associated with them.
If it's product of exponentials, then it's just an extra matrix multiple for pTc and it should only be for linked connected to actuated links.

One argument is that we should be computing the extra matrix multiple since a particular chain configuration could involve the end effector being connected by a fixed link and a user wants to get the final fixed link rather than jump through mental hoops to compute them. In that perspective we're more in alignment with Lynch and Park.

In the case of the A1, we only have 4 extra links connected by fixed joints that will be actuated. The rest are all connected directly to the base and shouldn't be involved in any kino-dynamic computations. Moreover you can now easily compute sensor reference frames wrt the body frame which someone like Jerred would actually appreciate.

Reply:
I don't even see an argument for having this lumping of fixed links in the first place. This is only in URDF files which is why the flag doesn't exist for SDF files, and I haven't seen a single example where this lumping is useful.

Sure there is an extra computational cost in PoE but that's the cost we pay for making GTDynamics far more user friendly.

Reply:
Elaborating on the previous point, if all our models were SDF files only, then would the topic of computational complexity and lumping the fixed joints still be relevant?

My answer is no, so why should it be the case with URDF files? Why should we see different behaviors for each file format?

Reply:
Case in point: https://github.com/osrf/gazebo/issues/2732

My guess is that the lumping was just for the purpose of gazebo simulation. The rationale isn't documented anywhere so it's just hearsay at this point.

Reply:
**Edit**: I just realized SdfConfig is only exposed in the `internal` namespace anyway so adding an option wouldn't really help, and propogating the option all the way up would be somewhat messy (exposing user to sdf:: doesn't seem the right direction).  Approving this PR, but I still think in the future there should be less computationally redundant ways to handle this.

---

> Where is the computational penalty coming from? If it's forward kinematics then fixed joints don't participate in that since they don't have any values associated with them. If it's product of exponentials, then it's just an extra matrix multiply for pTc and it should only be for links connected to actuated links.

I think fixed joints do have values associated with them, right?  They still have a joint angle key that is unconstrained and does nothing?  And all the corresponding factors too, so the factor graph increases in size by a decent bit?  Correct me if I'm wrong, but I'm fairly confident that this fixed joint will have all the same computations (including in "querying" and in factor graph solving) as a normal joint; it's just that the jacobians wrt joint angles will always be 0.

> The main concern is that you're thinking of what-ifs rather than what is necessary. If there is ever a use-case or need to have fixed joint lumping (which I highly doubt) then that person can create a test and make the appropriate change.

True, but I think I was thinking more in terms of _this_ is _changing_ the behavior to something less computationally efficient and it might be nice to leave the option in to _keep the existing_ more efficient behavior.  Really my main gripe is that these 2 lines make it impossible to not preserve fixed joints, when it would be just as reasonable to either make this behavior either a change-able default or a set-able non-default.

Reply:
I'll add the flag and then merge this in.